File: algorithm/leetcode124_binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxanswer = INT_MIN;
        maxsum(root, maxanswer);
        return maxanswer;
    }
    int maxsum(TreeNode* root, int &maxanswer) 
    {
    	if(root == NULL)
    		return 0;
    	int left = maxsum(root->left,maxanswer);
    	int right = maxsum(root->right,maxanswer);
    	int curmax = root->val;
    	if(left > 0)
        	curmax += left;
        if(right > 0)
        	curmax += right;
        if(curmax > maxanswer)
    		maxanswer = curmax;
    	return max(root->val,max(left+root->val,right+root->val));
    }
};

File: algorithm/Python3_version/leetcode124.binary-tree-maximum-path-sum.py
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.maxsum = float('-inf')
        def dfs(root):
            if not root:
                return 0
            l = dfs(root.left)
            r = dfs(root.right)
            v = root.val
            curmax = v
            if l>0:
                curmax += l
            if r>0:
                curmax += r
            self.maxsum = max(curmax, self.maxsum)
            return max(v, max(v+l, v+r))
        dfs(root)
        return self.maxsum